Improvements on ws_client

mbohlool opened the pull request:
The ws_client should be able to return an object so the caller can interact with server. This will happen if _preload_content is not True. In that case, the client will return an object that can interact with stdin/stdout/stderr of the remote container.

fixes #58

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
+
+    # TODO: This method does not seem to work.
+    def write_stdin(self, data):
+        self.sock.send(data)
```
Is the connection still open to even send the data? How does the underlining api url call look like when you try to tack on more `exec` commands to an already successful call?

mbohlool replied:
The connection stays open until explicitly closed by either side. My example code does not close the connection before sending data.

mrmcmuffinz commented on examples/exec.py:
```diff
@@ -0,0 +1,94 @@
+        print("STDERR: %s" % resp.read_stderr())
+
+exit(1)
+# This part does not work yet. resp.write_stdin does not work.
```
Could it be that you need to set the websocket to lingering, so that the connection stays open on both ends till the action you are trying to execute completes and forwards back the result?

mbohlool replied:
In the example bellow, I only run sh without closing the connection. _preload_content is False so the call will stays open. Then on line 94 inside the loop I send commands one by one. This example code will not end by itself because the connection stays open.

mbohlool commented:
@mrmcmuffinz @dims this is ready for review.

mrmcmuffinz commented:
If you don't mind I can review it tomorrow, I'm on the east coast and it's close to midnight here. I want to get some sleep.

mbohlool commented:
@mrmcmuffinz no rush. thanks.

dims commented on examples/exec.py:
```diff
@@ -0,0 +1,89 @@
+from kubernetes.client.rest import ApiException
+
+config.load_kube_config(
+    context="gke_cloud-kubernetes-dev_us-central1-f_mehdy-cluster")
```
have a more generic name here?

mbohlool replied:
Done.

dims commented:
@mbohlool : looks good, i haven't tried it yet. Though that reminds me that we should add a e2e test especially for the new code path where we pull data on demand.

mbohlool commented:
@dims. Will add an e2e test.

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -23,22 +25,17 @@
 
         # We just need to pass the Authorization, ignore all the other
         # http headers we get from the generated code
         if 'Authorization' in headers:
```
Do we need to be concerned that headers is None or that there is no `Authorization` in it?

mbohlool replied:
if check to see if there is authorization in it. if there is none, we don't need to pass any header. Fixed the case that headers is None.

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
+        self.update()
+        return self._stderr
+
+    def read_stderr(self):
```
Where was this in the example?

mbohlool replied:
end of the example, I've used this method.

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
+        self._stderr = ""
+        return ret
+
+    def read_all(self):
```
Where was this in the example?

mbohlool replied:
it used later in this file in GET method for non-interactive case.

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -12,9 +12,11 @@
+import time
 import collections
-import websocket
+from websocket import WebSocket, ABNF, enableTrace
```
Just a suggestion however can we organize these, `import` first then `from`?

mbohlool replied:
we have pep8 and sort_import running automatically and checking these. We cannot organize them the way we want unless we want to remove those checks.

mrmcmuffinz replied:
I see.

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
+    def is_open(self):
+        return self._connected
+
+    # TODO: This method does not seem to work.
```
If this does not work, do it make sense to keep it in the code or remove it?

mbohlool replied:
old comment. removed.

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
+            if op_code == ABNF.OPCODE_CLOSE:
+                self._connected = False
+                return
+            elif op_code == ABNF.OPCODE_BINARY or op_code == ABNF.OPCODE_TEXT:
```
Is there any other `op_code` that needs to be considered other than the ones you caught here and above?

mbohlool replied:
I don't think so.

mrmcmuffinz replied:
I don't know much about sockets however I see that there are more codes https://github.com/websocket-client/websocket-client/blob/master/websocket/_abnf.py#L102 where I thought we may have to cover our bases but I would like to leave that up to someone else who is more experienced in the area.

mbohlool replied:
I've modeled this from WebSocketApp here: https://github.com/websocket-client/websocket-client/blob/master/websocket/_app.py#L205. Ping and Pong are optional (and we can support it later). Cont can be ignored and it is only useful if you want to get notified on an incomplete frame. however, we may improve this later by doing something on "Cont" (like running update again).

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
 
-def GET(configuration, url, query_params, _request_timeout, headers):
+def GET(configuration, url, query_params, _request_timeout, _preload_content,
+        headers):
```
Documentation?

mbohlool replied:
GET is internal, but above methods need docs that will be added.

mrmcmuffinz commented on kubernetes/client/ws_client.py:
```diff
@@ -52,30 +49,87 @@ def __init__(self, configuration, url, headers):
+                        self._stdout += data
+
+    def run_forever(self, timeout=None):
+        if timeout:
```
Can we get an explanation on what run_forever means and are there any gotchas(what are the assumptions you are making)?

mbohlool replied:
will add documentation.

mbohlool commented on examples/exec.py:
```diff
@@ -0,0 +1,89 @@
+                                           command=exec_command,
+                                           stderr=True, stdin=False,
+                                           stdout=True, tty=False)
+print "Response: " + resp
```
There are many cases that we can cover. We can add more examples later (and feel free to add more).

mbohlool commented on examples/exec.py:
```diff
@@ -0,0 +1,89 @@
+exec_command = ['/bin/sh']
+resp = api.connect_get_namespaced_pod_exec(name, 'default',
+                                           command=exec_command,
+                                           stderr=True, stdin=True,
```
1. stdin is part of auto-generated arguments that we never touch.
2. stdin is working now :)

mrmcmuffinz replied:
Awesome!

mbohlool commented on kubernetes/client/ws_client.py:
```diff
@@ -23,22 +25,17 @@
 
 class WSClient:
     def __init__(self, configuration, url, headers):
```
I will add some documentation for methods in this class. __init__ specifically does not need docs in my opinion as this class is only created inside this file.

mbohlool commented:
@mrmcmuffinz, @dims addressed comments and also make the implementation more general to support any channel number (0,1,2 for std in,out,err). We may want to reuse this again in port forwarding api calls (they use websocket too). PTAL.

mrmcmuffinz commented:
General comment, was your intention to expose the channel constants for the end user like myself or just for the purpose of the websocket client?

mbohlool commented:
@mrmcmuffinz Channel numbering is a kubernetes thing as far as I can tell and I think it is not bad to let user know about it. But the helper function like read_stdout can hide that if user don't want to care about them. The whole channels expose can be useful later for port forwarding api implementation.

chekolyn commented on kubernetes/client/ws_client.py:
```diff
@@ -12,33 +12,40 @@
-                                         on_close=self.on_close,
-                                         header=[header] if header else None)
-        self.ws.on_open = self.on_open
+        if headers and 'Authorization' in headers:
```
I'm running into a small issue when running the exec.py example using header authentication. 
Looks like is a small case condition; it appears that the header key is "authorization" instead of "Authorization" and then we pass an empty header to the WS call.
```
(kubernetes-local) PAOC02S79NJG8WP:examples sa$ python exec.py
{'Content-Type': 'application/json', 'authorization': 'Bearer CLSMM1PCEbKNOgy2fz5runETCrf-yzV1ThPXAiJ4aLw', 'Accept': '*/*', 'User-Agent': 'Swagger-Codegen/1.0.0-snapshot/python'}
Traceback (most recent call last):
  File "exec.py", line 61, in <module>
    stdout=True, tty=False)
  File "/Users/sagui10/git/kubernetes/client-python/kubernetes/client/apis/core_v1_api.py", line 907, in connect_get_namespaced_pod_exec
    (data) = self.connect_get_namespaced_pod_exec_with_http_info(name, namespace, **kwargs)
  File "/Users/sagui10/git/kubernetes/client-python/kubernetes/client/apis/core_v1_api.py", line 1012, in connect_get_namespaced_pod_exec_with_http_info
    collection_formats=collection_formats)
  File "/Users/sagui10/git/kubernetes/client-python/kubernetes/client/api_client.py", line 329, in call_api
    _return_http_data_only, collection_formats, _preload_content, _request_timeout)
  File "/Users/sagui10/git/kubernetes/client-python/kubernetes/client/api_client.py", line 153, in __call_api
    _request_timeout=_request_timeout)
  File "/Users/sagui10/git/kubernetes/client-python/kubernetes/client/api_client.py", line 355, in request
    headers=headers)
  File "/Users/sagui10/git/kubernetes/client-python/kubernetes/client/ws_client.py", line 239, in websocket_call
    raise ApiException(status=0, reason=str(e))
kubernetes.client.rest.ApiException: (0)
Reason: Handshake status 403
```

> I added this to debug and make it work:
> 
>         # We just need to pass the Authorization, ignore all the other
>         # http headers we get from the generated code
>         if headers and 'Authorization' in headers:
>             header.append("Authorization: %s" % headers['Authorization'])
>         elif headers and 'authorization' in headers:
>             header.append("Authorization: %s" % headers['authorization'])
> 
>         print headers

mrmcmuffinz replied:
@chekolyn brings up a good point, however instead of doing the `elif headers and 'authorization' in headers:` we just need to do a case insensitive search instead. What do you think @mbohlool ?

mbohlool replied:
This should be always authorization with all lower case. I wonder why it worked for me before, I am investigating it. But I am agains making the check case insensitive as it will result in being inconsistent in different part of the code while we know this should be always lowecase.

mbohlool replied:
I've fixed this to be lower case authorization to be consistent with the rest of the client, also created #128 to further investigate this.

dims commented on kubernetes/client/ws_client.py:
```diff
@@ -12,33 +12,40 @@
-                                         on_close=self.on_close,
-                                         header=[header] if header else None)
-        self.ws.on_open = self.on_open
+        if headers and 'authorization' in headers:
```
Should we support both `Authorization` and `authorization`?

mbohlool replied:
I think we should be consistent. We used lowercase `authorization` everywhere else in our client. Based on the discussion in #128, I think we should have capital `Authorization` everywhere and case-insensitive checks. I will let another PR fixing #128 to do it consistently for whole client and keep this here to be consistent meanwhile.

dims commented:
One question inline, LGTM otherwise. thanks @mbohlool

mbohlool commented:
thanks @dims @mrmcmuffinz @chekolyn

mrmcmuffinz commented:
Hey @mbohlool to be clear I can not set in my pip requirements.txt `v1.0.0b2` for kubernetes client-python right?

mrmcmuffinz commented:
Btw, thanks for being patient and following through on the comments. Awesome work!

chekolyn commented:
Second that, awesome work!
@mbohlool with your consent I would like to port your WSClient implementation to [pykube](https://github.com/kelproject/pykube). Both projects have the same Apache License, would that be OK?

mbohlool commented:
@chekolyn That would be fine. But I think it is better for pykube to add a dependency on client-python. It could simplify their config and transport layers too (by delegating them to client-python) and get any patch/update for free.

mbohlool commented:
@mrmcmuffinz you're welcome. v1.0.0b2 has a dependency problem that @dims is fixing. Was your question related to that?
